File: 2024/Rust/src/day04.rs
use std::collections::HashMap;

use common::Vec2D;

type Parsed = HashMap<Vec2D<isize>, Letter>;
type Solution = u64;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Letter {
    X,
    M,
    A,
    S,
}

fn m_and_s(lhs: Letter, rhs: Letter) -> bool {
    match (lhs, rhs) {
        (Letter::M, Letter::S) => true,
        (Letter::S, Letter::M) => true,
        _ => false,
    }
}

pub fn parse(input: &str) -> Parsed {
    input
        .lines()
        .enumerate()
        .flat_map(|(y, line)| {
            line.chars()
                .enumerate()
                .map(move |(x, c)| ((x as isize, y as isize).into(), c))
        })
        .filter_map(|(pos, c)| match c {
            'X' => Some((pos, Letter::X)),
            'M' => Some((pos, Letter::M)),
            'A' => Some((pos, Letter::A)),
            'S' => Some((pos, Letter::S)),
            _ => None,
        })
        .collect()
}

const WORD: [Letter; 4] = [Letter::X, Letter::M, Letter::A, Letter::S];
const DIRECTIONS: [Vec2D<isize>; 8] = [
    Vec2D { x: 1, y: 0 },   // right
    Vec2D { x: -1, y: 0 },  // left
    Vec2D { x: 0, y: 1 },   // down
    Vec2D { x: 0, y: -1 },  // up
    Vec2D { x: 1, y: 1 },   // down-right
    Vec2D { x: -1, y: -1 }, // up-left
    Vec2D { x: 1, y: -1 },  // down-left
    Vec2D { x: -1, y: 1 },  // up-right
];

fn wordsearch(map: &Parsed) -> u64 {
    let mut count = 0;

    for (&coord, &letter) in map.iter() {
        if letter == *WORD.first().unwrap() {
            for direction in &DIRECTIONS {
                if check_word(map, coord, *direction) {
                    count += 1;
                }
            }
        }
    }

    count
}

fn check_word(map: &Parsed, start: Vec2D<isize>, direction: Vec2D<isize>) -> bool {
    let mut current_coord = start;
    for ch in WORD {
        if let Some(&map_ch) = map.get(&current_coord) {
            if map_ch != ch {
                return false;
            }
        } else {
            return false;
        }
        current_coord += direction;
    }
    true
}

pub fn part1(parsed: &Parsed) -> Solution {
    wordsearch(parsed)
}

pub fn part2(parsed: &Parsed) -> Solution {
    let mut count = 0;

    for (&coord, &letter) in parsed.iter() {
        if letter == Letter::A {
            // if top-left of us there is a letter, bottom right needs to be the other letter
            // if top-right of us there is a letter, bottom left needs to be the other letter

            let top_left = coord + (-1, -1).into();
            let top_right = coord + (1, -1).into();
            let bottom_left = coord + (-1, 1).into();
            let bottom_right = coord + (1, 1).into();

            if let Some(&top_left_letter) = parsed.get(&top_left) {
                if let Some(&bottom_right_letter) = parsed.get(&bottom_right) {
                    if let Some(&top_right_letter) = parsed.get(&top_right) {
                        if let Some(&bottom_left_letter) = parsed.get(&bottom_left) {
                            if m_and_s(top_left_letter, bottom_right_letter)
                                && m_and_s(top_right_letter, bottom_left_letter)
                            {
                                count += 1;
                            }
                        }
                    }
                }
            }
        }
    }

    count
}

pub fn run(input: &str) {
    let parsed = parse(input);
    println!("Part 1: {}", part1(&parsed));
    println!("Part 2: {}", part2(&parsed));
}

#[cfg(test)]
mod tests {
    use super::*;

    const INPUT: &str = "MMMSXXMASM
MSAMXMSMSA
AMXSXMAAMM
MSAMASMSMX
XMASAMXAMM
XXAMMXXAMA
SMSMSASXSS
SAXAMASAAA
MAMMMXMMMM
MXMXAXMASX
";

    const P1_SOLUTION: Solution = 18;
    const P2_SOLUTION: Solution = 9;

    #[test]
    fn test_part1() {
        assert_eq!(part1(&parse(INPUT)), P1_SOLUTION);
    }
    #[test]
    fn test_part2() {
        assert_eq!(part2(&parse(INPUT)), P2_SOLUTION);
    }
}

File: 2024/Haskell/src/Day04.hs
module Day04
  ( day04,
  )
where

import AoC
import Control.Monad (guard)
import Data.HashMap.Strict qualified as HM
import Data.Hashable (Hashable (hash, hashWithSalt))
import Data.Maybe (fromMaybe)
import Data.Text (Text)
import Data.Text qualified as Text
import Data.Vector.Internal.Check (check)

type Parsed = HM.HashMap Vec2D Char

type Solution = Int

data Vec2D = Vec2D
  { x :: Int,
    y :: Int
  }
  deriving (Show, Eq)

instance Hashable Vec2D where
  hashWithSalt salt (Vec2D x y) = salt `hashWithSalt` (476743 * hash x + 152857 * hash y)

instance Num Vec2D where
  (Vec2D x1 y1) + (Vec2D x2 y2) = Vec2D (x1 + x2) (y1 + y2)
  (Vec2D x1 y1) - (Vec2D x2 y2) = Vec2D (x1 - x2) (y1 - y2)
  (Vec2D x1 y1) * (Vec2D x2 y2) = Vec2D (x1 * x2) (y1 * y2)
  abs (Vec2D x y) = Vec2D (abs x) (abs y)
  signum (Vec2D x y) = Vec2D (signum x) (signum y)
  fromInteger n = Vec2D (fromInteger n) (fromInteger n)

-- Constants

word :: String
word = "XMAS"

directions :: [Vec2D]
directions =
  [ Vec2D 0 1, -- up
    Vec2D 0 (-1), -- down
    Vec2D 1 0, -- right
    Vec2D (-1) 0, -- left
    Vec2D 1 1, -- up-right
    Vec2D (-1) 1, -- up-left
    Vec2D 1 (-1), -- down-right
    Vec2D (-1) (-1) -- down-left
  ]

corners :: [Vec2D]
corners =
  [ Vec2D (1) (-1), -- top right
    Vec2D (1) (1), -- top left
    Vec2D (-1) (-1), -- bottom right
    Vec2D (-1) (1) -- bottom left
  ]

checkWord :: Parsed -> Vec2D -> Vec2D -> Bool
checkWord map current' direction = go word current'
  where
    go (ch : xs) current = case map HM.!? current of
      Just mapCh | mapCh == ch -> go xs (current + direction)
      _ -> False
    go [] _ = True

p1 :: Parsed -> Int
p1 map = foldl' f 0 (HM.toList map)
  where
  first = head word
  f :: Int -> (Vec2D, Char) -> Int
  f acc (coord, letter) =
    if letter == first
      then acc + (length $ filter (checkWord map coord) directions)
      else acc

p2 map = length $ filter f $ HM.toList map
  where
    m_and_s 'M' 'S' = True
    m_and_s 'S' 'M' = True
    m_and_s _ _ = False

    g coord letter = do
      guard $ letter == 'A'
      let directions = (coord +) <$> corners
      [tl, tr, bl, br] <- (map HM.!?) `traverse` directions
      return $ m_and_s tl br && m_and_s tr bl

    f (coord, letter) = fromMaybe False (g coord letter)

parse :: Text -> HM.HashMap Vec2D Char
parse input =
  let lines = Text.unpack <$> Text.lines input
   in HM.fromList do
      (y, line) <- zip [0 ..] lines
      (x, ch) <- zip [0 ..] line
      return (Vec2D x y, ch)

day04 :: Runner Parsed Solution
day04 =
  let year = 2024
      day = 04
      parser :: Text -> Maybe Parsed
      parser = return . parse
      part1 :: Parsed -> Maybe Solution
      part1 = return . p1
      part2 :: Parsed -> Maybe Solution
      part2 = return . p2
   in Runner {..}
